[Fix #1868] Only register an offense in Count for select with block or &:something format

rrosenblum opened the pull request:
This fixes #1868.

This was a bit more difficult to fix than I thought it would be. I might be missing something, but in general, it has proven challenging to properly parse method chains that can take parameters or blocks. This was similar to the challenges in `Performance/Sample`. 

The only parameters to `select`, that will cause an offense, are the `&:something` format.

The other notable functionality to this cop, is that if a block is passed to both `select` and `count`, it will not register an offense. This functionality is present in version `0.31.0`. This will miss some offenses on `Enumerable`, but should be safe implementation with Rails.

jonas054 commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@
 ### Bugs fixed
 
 * Don't count required keyword args when specifying `CountKeywordArgs: false` for `ParameterLists`. ([@sumeet][])
+* [#1868](https://github.com/bbatsov/rubocop/issues/1868): Do not register an offense in Performance/Count when `select` is called with symbols or strings as the parameters. ([@rrosenblum][])
```
Perhaps you'd want to have backticks around Performance/Count.

rrosenblum replied:
Good idea, I will add them

jonas054 commented on lib/rubocop/cop/performance/count.rb:
```diff
@@ -27,56 +30,67 @@ class Count < Cop
 
-          [expression, first_method, second_method, third_method]
+        def contains_selector?(node)
+          node.respond_to?(:loc) && node.loc.respond_to?(:selector)
```
A node that doesn't respond to `loc` sounds strange, and indeed, the left hand side operator can be removed without any spec failures.

rrosenblum replied:
I will try to find the specific use case for why I added this in. I think it is to handle `nil` and Symbols. When extracting variables using `left, right = *node` and then using the same style to expand `left` and `right`, elements that wind up being methods are Symbols, not Nodes.

I will find the exact use case and respond back.

rrosenblum replied:
The use case for this came from running this cop again the RuboCop source code.

```ruby
        def check_duplicate_methods(names)
          dups = names.each_with_object({}) do |item, accum|
            accum[item[0]] ||= []
            accum[item[0]] << item[1]
          end 

          dups.each do |method, nodes|
            next if nodes.size < 2 
            lines = nodes.map { |node| node.loc.line }

            add_offense(nodes.last,
                        :keyword,
                        format(MSG, method, lines.join(', ')))
          end 
        end
```
In this code, the line `next if nodes.size < 2 ` would cause `undefined method `loc' for :nodes:Symbol`. More specifically, the node being processed was `nodes.size`.

I have simplified down the source that causes this error, and will be adding a test case that covers this

jonas054 commented on lib/rubocop/cop/performance/count.rb:
```diff
@@ -27,56 +30,67 @@ class Count < Cop
+          expression, selector, = parse(node)
 
-          return if first_method == :reject || second_method == :reject
+          selector_loc = if selector.is_a?(Symbol)
```
Perhaps better to call it `selector_range`, given its type.

jonas054 commented on lib/rubocop/cop/performance/count.rb:
```diff
@@ -27,56 +30,67 @@ class Count < Cop
-          begin_pos = if SELECTORS.include?(first_method)
-                        return if second_method.is_a?(Symbol)
-                        expression.loc.selector.begin_pos
+          begin_pos = if selector.is_a?(Symbol)
```
There's some duplication here. Perhaps this and the `selector_loc` algorithm can be extracted into a common method?

jonas054 commented on lib/rubocop/cop/performance/count.rb:
```diff
@@ -27,56 +30,67 @@ class Count < Cop
-                        return if second_method.is_a?(Symbol)
-                        expression.loc.selector.begin_pos
+          begin_pos = if selector.is_a?(Symbol)
+                        expression.parent.loc.selector.begin_pos if expression
```
There are no spec examples that make `expression` become `nil`.

jonas054 commented on lib/rubocop/cop/performance/count.rb:
```diff
@@ -27,56 +30,67 @@ class Count < Cop
 
+          begin_pos ||= node.loc.expression.begin_pos
+
           range = Parser::Source::Range.new(node.loc.expression.source_buffer,
```
I'd like to see a test on `highlights` in the specs so we know that the range is correct (while you're at it, same problem exists on `master`).

rrosenblum replied:
Good idea. I have caused the wrong, or poor highlighting in the past.

In general, I think that it would be a good idea for all cops to start have tests for `highlights`.

rrosenblum commented:
I have updated the code. Some of the logic has been simplified down. I have added some test cases to show that the extra checks in `parse` are needed for parsing unrelated code. I have also added tests for `highlights`.

jonas054 commented:
:+1: Looks good!

bbatsov commented:
:+1:
